controllers/openstackdataplanenodeset_controller.go

```diff
@@ -257,45 +230,56 @@ func (r *OpenStackDataPlaneRoleReconciler) Reconcile(ctx context.Context, req ct
 
 	// Reconcile BaremetalSet if required
 	if len(instance.Spec.BaremetalSetTemplate.BaremetalHosts) > 0 {
-		// Reset the RoleBareMetalProvisionReadyCondition to unknown
-		instance.Status.Conditions.MarkUnknown(dataplanev1.RoleBareMetalProvisionReadyCondition,
+		// Reset the NodeSetBareMetalProvisionReadyCondition to unknown
+		instance.Status.Conditions.MarkUnknown(dataplanev1.NodeSetBareMetalProvisionReadyCondition,
 			condition.InitReason, condition.InitReason)
 		isReady, err := deployment.DeployBaremetalSet(ctx, helper, instance,
-			nodes, allIPSets, dnsAddresses)
+			allIPSets, dnsAddresses)
 		if err != nil || !isReady {
 			return ctrl.Result{}, err
 		}
 	} else {
-		instance.Status.Conditions.Remove(dataplanev1.RoleBareMetalProvisionReadyCondition)
-	}
-
-	// Ensure all nodes are in SetupReady state
-	setupReadyNodes := 0
-	for _, node := range nodes.Items {
-		if node.IsSetupReady() {
-			setupReadyNodes++
-		}
+		instance.Status.Conditions.Remove(dataplanev1.NodeSetBareMetalProvisionReadyCondition)
 	}
 
-	if setupReadyNodes < len(nodes.Items) {
-		return ctrl.Result{}, err
-	}
-
-	// Generate Role Inventory
-	roleConfigMap, err := deployment.GenerateRoleInventory(ctx, helper, instance,
-		nodes.Items, allIPSets, dnsAddresses)
+	// Generate NodeSet Inventory
+	roleConfigMap, err := deployment.GenerateNodeSetInventory(ctx, helper, instance,
+		allIPSets, dnsAddresses)
 	if err != nil {
 		util.LogErrorForObject(helper, err, fmt.Sprintf("Unable to generate inventory for %s", instance.Name), instance)
 		return ctrl.Result{}, err
 	}
 
+	// Verify Ansible SSH Secret
+	ansibleSSHPrivateKeySecret = instance.Spec.NodeTemplate.AnsibleSSHPrivateKeySecret
+	if ansibleSSHPrivateKeySecret != "" {
+		_, result, err = secret.VerifySecret(
+			ctx,
+			types.NamespacedName{Namespace: instance.Namespace, Name: ansibleSSHPrivateKeySecret},
+			[]string{
+				"ssh-privatekey",
+			},
+			r.Client,
+			time.Duration(5)*time.Second,
+		)
+		if err != nil {
+			return result, err
+		}
+	}
+
 	// all setup tasks complete, mark SetupReadyCondition True
 	instance.Status.Conditions.MarkTrue(dataplanev1.SetupReadyCondition, condition.ReadyMessage)
 
-	r.Log.Info("Role", "DeployStrategy", instance.Spec.DeployStrategy.Deploy,
-		"Role.Namespace", instance.Namespace, "Role.Name", instance.Name)
-	if instance.Spec.DeployStrategy.Deploy {
-		r.Log.Info("Starting DataPlaneRole deploy")
+	// The OpenStackDataPlane controller will annotate the nodeSet when it's ready to perform
+	// the executions. We need to check for the presense of this annotation.
+	_, deployAnnotation := instance.ObjectMeta.Annotations[edpmDeployAnnotationName]
```

Comment:
Let's drop the annotation from this PR. It is unrelated.

Reply:
We can, but it's the trigger from the `OpenStackDataPlane` Controller to run the rest of this deployment. Else we're back to the state where the OpenStackDataPlane doesn't do anything?

Reply:
I think what you probably would like me to do is to use the `OpenStackDataPlane` controller to update the `nodeSet.Spec.DeployStrategy.Deploy = true` as the trigger for the deployment. But that seems redundant. If a user creates a `NodeSet` with `deploy: true` it's going to run the entire deployment anyway. 

That makes `OpenStackDataPlane` feel like a pretty unnecessary abstraction where we need to maintain code to search for nodesets by labels, etc just so that we can do something that the user might have already done anyway. In that scenario, it feels like we might as well just drop the controller and save the space in `etcd`. Even in it's current state, it is still a lot of code to maintain for basically no gain.

So, I guess the summation of my point is that, if what we want is the ability to define infrastructure independent of the ability to execute Ansible playbooks against said infrastructure. I strongly feel that we need a trigger outside of the user controlled spec to achieve that. Otherwise it creates a potentially confusing interface for users who can still do things in multiple different ways and might be left wondering if they're doing it right. Either removing it, or making it functionally necessary would in my opinion be more intuitive for users.

Reply:
As it relates to this PR, I'd just like to separate out the changes related to DeployStrategy.Deploy. I think we said that in the design doc.

Reply:
No worries, I'll remove it.

Just to clarify though, my intention wasn't to change the `deployStrategy.Deploy`. Definitely agree to leave that. One of the very first things I called out in the design doc was that having multiple different controllers presented users with multiple different ways to arrive at the same outcome. Consequently, we need to write code, tests, CI pipelines to ensure reliability of those methods. Additionally, it presents a possible point of confusion for users if they can do things a number of different ways that ultimately achieve the same outcome.

So my intention here wasn't to change the deployStrategy implementation - since we had agreed to leave it. My intention was to instead supplement that implementation with a mechanism that would enforce a workflow. Necessitating that users define a `nodeSet` then a `dataplane` object pushes all use cases towards something that is predictable and thus reduces the complexity - both perceived complexity by users and technical complexity of the code required to maintain it.

You have already called out that we don't need to necessarily arrive at the perfect location as part of this work, since it would most likely mean continual scope creep and never getting anywhere. So I'm happy to continue this as part of an iterative approach moving forward. 

For now, I'll remove the annotation and just leave the dataplane CR as a symbolic placeholder until we have a better job for it to do.